src/Workspaces/Core/Portable/Utilities/AsyncLazy`1.cs

```diff
@@ -530,24 +544,36 @@ private void CompleteFromTaskSynchronouslyStub(object task)
 
             public void CompleteFromTaskSynchronously(Task<T> task)
             {
-                if (task.Status == TaskStatus.RanToCompletion)
+                // AsyncTaskMethodBuilder doesn't give us Try* methods, and the Set methods may throw if the task
+                // is already completed. The belief is that the race is somewhere between rare to impossible, and
+                // so we'll do a quick check to see if the task is already completed or otherwise just give it a shot
+                // and catch it if it fails
+                if (_taskBuilder.Task.IsCompleted)
                 {
-                    if (TrySetResult(task.Result))
-                    {
-                        _cancellationTokenRegistration.Dispose();
-                    }
+                    return;
                 }
-                else if (task.Status == TaskStatus.Faulted)
+
+                try
                 {
-                    if (TrySetException(task.Exception))
+                    if (task.Status == TaskStatus.RanToCompletion)
```

Comment:
Task.Status needs to do a bunch of checks to figure out the right enumeration value to return.  In contrast, the Is* properties only need to do a single bit flag check to determine whether that is or is not the status.  I don't know whether this level of optimization matters here, so ignore me if it doesn't.  If it does, consider rewriting to:
```
if (task.IsFaulted)
{
    _taskBuilder.SetException(task.Exception);
}
else if (task.IsCanceled)
{
    CancelSynchronously();
}
else
{
    _taskBuilder.SetResult(task.Result);
}
```